sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala

```diff
@@ -1259,56 +1259,66 @@ private[lf] object SBuiltinFun {
-                          else
-                            checkContractUpgradable(coid, srcContract, dstContract) { () =>
-                              executeExpression(machine, SEPreventCatch(dstView)) { dstViewValue =>
-                                if (srcViewValue != dstViewValue) {
-                                  Control.Error(
-                                    IE.Upgrade(
-                                      IE.Upgrade.ViewMismatch(
-                                        coid,
-                                        interfaceId,
-                                        srcTplId,
-                                        dstTplId,
-                                        srcView = srcViewValue.toUnnormalizedValue,
-                                        dstView = dstViewValue.toUnnormalizedValue,
-                                      )
-                                    )
-                                  )
-                                } else
-                                  k(SAny(Ast.TTyCon(dstTplId), dstArg))
+      viewInterfaceSafe(machine, interfaceId, srcTplId, srcArg) { mbSrcView =>
+        resolvePackageName(machine, pkgName) { pkgId =>
+          val dstTplId = srcTplId.copy(packageId = pkgId)
+          machine.ensurePackageIsLoaded(
+            dstTplId.packageId,
+            language.Reference.Template(dstTplId),
+          ) { () =>
+            ensureTemplateImplementsInterface(machine, interfaceId, coid, dstTplId) {
+              fromInterface(machine, srcTplId, srcArg, dstTplId) {
+                case None =>
+                  Control.Error(IE.WronglyTypedContract(coid, dstTplId, srcTplId))
+                case Some(dstArg) =>
+                  viewInterface(machine, interfaceId, dstTplId, dstArg) { dstView =>
+                    def withViewValue(mbSrcViewValue: Option[SValue]): Control[Question.Update] = {
+                      getContractInfo(
+                        machine,
+                        coid,
+                        dstTplId,
+                        dstArg,
+                        allowCatchingContractInfoErrors = false,
+                      ) { dstContract =>
+                        // If the destination and src templates are the same, we skip the computation
+                        // of the destination template's view and the validation of the contract info.
+                        if (dstTplId == srcTplId)
+                          k(SAny(Ast.TTyCon(dstTplId), dstArg))
+                        else
+                          checkContractUpgradable(coid, srcContract, dstContract) { () =>
+                            executeExpression(machine, SEPreventCatch(dstView)) { dstViewValue =>
+                              mbSrcViewValue match {
+                                case None => ()
+                                case Some(srcViewValue) =>
+                                  if (srcViewValue != dstViewValue) {
```

Comment:
Why are we checking the view values to match here?

Isn't that what we wanted to remove?

Reply:
It still emits a warning if the view interface changed - we're sure we don't want any indication? At INFO loglevel, I can see this helping devs who've accidentally changed their view expression, while providing comparatively little noise for devs who did it on purpose.

Reply:
How do you even compute the source view if you don't have the creating package available anymore?

The ROI seems low to me. The are many ways how an upgrade can go wrong. The main protection against mistakes are reviews of the diff of the source-code. That one should work well for a wrong interface view.

And if it does go wrong, then the upgrade can be rolled-back and a fixed version can be deployed.

Reply:
Essentially, I was thinking about large refactors where it isn't immediately easy to confirm that a view is always going to change consistently, especially early in the development cycle before you have a source code audit. I'd like engineers to be able to verify and have "sanity checks" during development and after something goes seriously wrong - DEBUG seemed like an appropriate log level for that.

After a bit more thought, however, I think it's best to let clients implement that themselves if they want to - after all, that's not too difficult, you just query your target test contract in V1 and V2 and then compare them.

I'll go ahead and remove this :+1:

Reply:
d36d770882